felix/dissimilarity.py

```diff
@@ -0,0 +1,188 @@
+import pickle
+from gerrypy.data.load import *
+from gerrypy.analyze.viz import *
+import numpy as np
+import pandas as pd
+import itertools
+
+# Main Functions
+def k_most_dissimilar(plans, num_blocks, k):
+    """ Given a set of plans represented as a list of list of lists,
+    returns the k most dissimilar plans
+    Args:
+        plans: a set of maps
+        num_blocks: the number of blocks in the state
+        k: the number of most distant maps to return
+    """
+    maps = convert_plans(plans, num_blocks)
+    dist_matrix = compute_dist_matrix(maps)
+    selected_maps = furthest_maps(dist_matrix, maps, k)
+    return convert_back(selected_maps)
+
+
+# Helper Functions
+def convert_plans(plans, num_blocks):
+    """
+    Converts plans from a list of maps, each of which contain a list of districts, each of
+    which contains a list of blocks in that district into a 3D binary numpy tensor,
+    with the first dimension representing the number of maps, the second dimension representing
+    the number of districts, and the third dimension a binary array of length (num_blocks),
+    where a 1 in the ith entry indicates that the district contains block i.
+    Args:
+        plans: the list of maps to convert
+        num_blocks: the largest number of blocks in a single district. Used for the conversion
+        to binary arrays.
+    """
+    possible_districts = pd.Series(np.arange(0, num_blocks))
+    converted_plans = []
+    for i in range(len(plans)):
+        plan = plans[i]
+        converted_plan = []
+        for j in range(len(plan)):
+            district = plan[j]
+            converted_plan.append(possible_districts.isin(district))
+        converted_plans.append(converted_plan)
+    return np.array(converted_plans).astype(int)
+
+
+def compute_dist_matrix(maps):
```

Comment:
Implementation suggestion: This might be faster (no loops). You could omit the line adding the symmetry if you don't care about symmetry, or if you are careful enough to only use indices for `i >= j`
```suggestion
def compute_dist_matrix(maps):
    # Compute matrix multiplication between each pair of maps, transpose it to (maps, maps, districts, blocks)
    prod = np.tensordot(maps, maps.T, axes=1).transpose(0, 3, 1, 2) 
    dist = 1000 * (1/np.mean(np.max(prod, axis=-1), axis=-1)) 
    # To make it symmetric 
    dist[np.triu_indices(dist.shape[0], 1)] = dist.T[np.triu_indices(dist.shape[0], 1)] 
    return dist 
```